### ISSUE TITLE
Is Guest `compute_loss` in `hetero lr ` calculated right?

### ISSUE THREAD

**Issue by gxcuit** [FederatedAI/FATE#3596]
**Is your feature request related to a problem? Please describe.**

I noticed that in fate 1.7.0, the `compute_loss` of Guest in `hetero-lr` has updated. However, I think the code/logic can be optimized.  
The format of  loss `loss = (1/N)*∑(log2 - 1/2*ywx + 1/8*(wx)^2), where y is label, w is model weight and x is features` **is  under the assumption that the label is either -1 or 1.** (Please kindly refer ref[1]). Therefore, when computing `ywx`, it is not necessary to judge `y==1 or not`.

https://github.com/FederatedAI/FATE/blob/6a9c2dd6ff95fcab6f336a2e188a0c58f3777d39/python/federatedml/optim/gradient/hetero_lr_gradient_and_loss.py#L77-L91

**Describe the solution you'd like**
Optimize the logic of calculating loss.

**Describe alternatives you've considered**
I have already submitted a PR. Please kindly refer to #3599 .

NOTE: This update does not change the result. The flowing test are based on the  examples [hetero_lr_normal_conf.json](https://github.com/FederatedAI/FATE/blob/6a9c2dd6ff95fcab6f336a2e188a0c58f3777d39/examples/dsl/v2/hetero_logistic_regression/hetero_lr_normal_conf.json) and  [hetero_lr_normal_dsl.json](https://github.com/FederatedAI/FATE/blob/6a9c2dd6ff95fcab6f336a2e188a0c58f3777d39/examples/dsl/v2/hetero_logistic_regression/hetero_lr_normal_dsljson).

Before optimization, the loss is :
```
[INFO] [2022-01-05 10:07:18,599] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 0,  loss:0.6931471805599453, is_converged: False
[INFO] [2022-01-05 10:07:28,858] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 1,  loss:50.081458179851246, is_converged: False
[INFO] [2022-01-05 10:07:38,926] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 2,  loss:3.4059021475555524, is_converged: False
[INFO] [2022-01-05 10:07:49,056] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 3,  loss:0.8378388622788911, is_converged: False
[INFO] [2022-01-05 10:07:59,385] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 4,  loss:0.4590097444373968, is_converged: False
[INFO] [2022-01-05 10:08:09,617] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 5,  loss:0.38025476990824936, is_converged: False
[INFO] [2022-01-05 10:08:19,102] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 6,  loss:0.36010687727146123, is_converged: False
[INFO] [2022-01-05 10:08:29,389] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 7,  loss:0.3531603116504674, is_converged: False
[INFO] [2022-01-05 10:08:40,491] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 8,  loss:0.34931024599470156, is_converged: False
[INFO] [2022-01-05 10:08:51,619] [202201051005288375370] [2428:139915861284672] - [base_linear_model_arbiter.fit] [line:129]: iter: 9,  loss:0.34617299941223556, is_converged: False
```

After optimization, the loss is : 
```
[INFO] [2022-01-05 11:04:01,939] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 0,  loss:0.6931471805599453, is_converged: False
[INFO] [2022-01-05 11:04:06,874] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 1,  loss:50.081458179851246, is_converged: False
[INFO] [2022-01-05 11:04:11,948] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 2,  loss:3.4059021475555524, is_converged: False
[INFO] [2022-01-05 11:04:16,346] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 3,  loss:0.8378388622788911, is_converged: False
[INFO] [2022-01-05 11:04:22,247] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 4,  loss:0.45900974443739684, is_converged: False
[INFO] [2022-01-05 11:04:27,171] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 5,  loss:0.38025476990824936, is_converged: False
[INFO] [2022-01-05 11:04:33,313] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 6,  loss:0.36010687727146123, is_converged: False
[INFO] [2022-01-05 11:04:38,238] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 7,  loss:0.35316031165046735, is_converged: False
[INFO] [2022-01-05 11:04:43,996] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 8,  loss:0.34931024599470156, is_converged: False
[INFO] [2022-01-05 11:04:48,494] [202201051103012517110] [27002:139957631616832] - [base_linear_model_arbiter.fit] [line:129]: iter: 9,  loss:0.34617299941223556, is_converged: False
```
The above-mentioned log can be found in the `arbiter/DEBUG.log`

**Additional context**
REF[1] The paper 
https://github.com/FederatedAI/FATE/blob/6a9c2dd6ff95fcab6f336a2e188a0c58f3777d39/doc/federatedml_component/logistic_regression.md?plain=1#L46-L47
![image](https://user-images.githubusercontent.com/18736196/148156442-a96c5748-bfc4-4ea4-89f7-f25279743cbb.png)

**Comment by dylan-fan** [comment]
paillier add operator is faster than scalar multiply. So , add this optim in 1.7.0

**Comment by gxcuit** [comment]
> applyPartitions

OK, I see. However, it brings the cost of two  `applyPartitions` operations. Does this make scene? How should we balance the computation cost between  `applyPartitions`  and paillier add operator?

**Comment by mgqa34** [comment]
@gxcuit Maybe you can refer to https://github.com/FederatedAI/FATE/blob/v1.6.1/python/federatedml/optim/gradient/hetero_lr_gradient_and_loss.py#L78, 
in older version, fate's developer use the same strategy as your pr. 
But in fate-v1.7.0, the logic changes due to two reason:
    (1) paillier add operator is faster than scalar multiply
    (2) the code in your pr, will reproduce two encrypted table, which is more bigger.

**Comment by gxcuit** [comment]
@mgqa34 
Thanks for your reply. Now I know that fate-v1.7.0 is much more efficient, and my previous PR does not make scene.

**I still have one question.**
In https://github.com/FederatedAI/FATE/blob/2c39286a30d1abd6c6ca38fcda51761631366666/python/federatedml/optim/gradient/hetero_lr_gradient_and_loss.py#L99-L101

**`host_forward got from host is 0.25 * wx_host`, instead of `wx_host`**

So, `wxg_wxh = 0.25 * wx_host * wx_guest`

Therefore, is  `2 * wxg_wxh` calculated right? I think it should be `8* wxg_wxh`, just as I PR in #3599 

Looking for your reply. @mgqa34  @dylan-fan

**Comment by nemirorox** [comment]
Hi,
Thank you for pointing out the error and submitting PR. You are right on the loss calculation. We will check and merge the pr soon.
